tutorial_buggy_impls.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * CONTAINS BUGGY IMPLEMENTATIONS FOR tutorial.ccp
+ *
+ * You can turn on buggy solutions by compiling with the following
+ * preprocessor symbols defined.
+ *
+ * SLIDE_RIGHT_BUG_1
+ * The implementation slides left instead of right.
+ * 
+ * SLIDE_RIGHT_BUG_2
+ * The implementation doesn't copy the last element to the beginning.
+ * 
+ * SLIDE_RIGHT_BUG_3
+ * 
+ * FLIP_BUG_1
+ * 
+ * FLIP_BUG_2
+ * 
+ * FLIP_BUG_3
+ * 
+ * 
+ */
+
+#include "tutorial.hpp"
+#include <iostream>
+
+using namespace std;
+
+// MODIFIES: the elements in v
+// EFFECTS:  All elements are "shifted" right by one unit, with the
+//           last element wrapping around to the beginning.
+// EXAMPLE:  If v contains [0,1,3,3,4], it would be modified to
+//           contain [4,0,1,3,3]
+
+#if defined SLIDE_RIGHT_BUG_1
+// Bug Description: The implementation slides left instead of right.
+
+void slideRight(std::vector<int> &v) {
+  int *arr = v.data();
+  int n = static_cast<int>(v.size());
+  int first = *arr;
+  for(int *ptr = arr+1; ptr < arr + n; ++ptr) {
+    *(ptr-1) = *ptr;
+  }
+  *(arr+n-1) = first;
+}
+
+#elif defined SLIDE_RIGHT_BUG_2
+// Bug Description: The implementation doesn't copy the last element to the beginning.
+
+void slideRight(std::vector<int> &v) {
+  int *arr = v.data();
+  int n = static_cast<int>(v.size());
+  for (int* p = arr+n-1; arr < p; --p)
+    *p = *(p-1);
+}
+
+#elif defined SLIDE_RIGHT_BUG_3
+// Bug Description: The implementation fails for an empty vector
+
+void slideRight(std::vector<int> &v) {
+  if (v.empty()) { v[0] = v[100000]; return; }
+  int *arr = v.data();
+  int n = static_cast<int>(v.size());
+  int last = *(arr+n-1);
+  for (int* p = arr+n-1; arr < p; --p)
+    *p = *(p-1);
+  *arr = last;
+}
+
+#else
+
+void slideRight(std::vector<int> &v) {
```

Comment:
This is broken on empty vectors...

Reply:
Whoops. Thanks for fixing this. Also, I like using the naturally broken one for bug 3 (enabled by the deterministic failure of .at()) rather than my hacky one.